agent/consul/streaming_endpoint.go

```diff
@@ -29,37 +29,49 @@ func (h *ConsulGRPCAdapter) Subscribe(req *stream.SubscribeRequest, server strea
 	snapshotCh := make(chan stream.Event, 32)
 	go snapshotFunc(server.Context(), snapshotCh, req.Key)
 
+	// Resolve the token and create the ACL filter.
+	rule, err := h.srv.ResolveToken(req.Token)
+	if err != nil {
+		return err
+	}
+	filt := newACLFilter(rule, h.srv.logger, h.srv.config.ACLEnforceVersion8)
+
 	// Wait for the events to come in and forward them to the client.
-	var snapshotDone bool
-	for !snapshotDone {
-		select {
-		case <-server.Context().Done():
-			return nil
-		case event, ok := <-snapshotCh:
-			h.srv.logger.Printf("[INFO] consul: sent snapshot event to key %q", req.Key)
-			if !ok {
-				h.srv.logger.Printf("[INFO] consul: finished sending snapshot")
-				snapshotDone = true
-				break
-			}
-			if err := server.Send(&event); err != nil {
-				return err
-			}
+	for event := range snapshotCh {
+		event.SetACLRules()
```

Comment:
Can we just have `allowEvent` call this? Then we don't have to remember to prime it everywhere we might filter an event to ensure it's set properly?

It would be kinda nice if there was an easy way to have protoc generate a hard coded method into each event type that just returns the correct set somehow rather than need an extra method but I think that's probably a complication we don't need to work on just now.

Reply:
Realised that doesn't work because then calling `filter.applyEvent` on a client would override the required ACLs with the ones the client code knew about not the ones the server sent on the wire.

🤔 There is something a little bit gross about being able to accidentally do this. I wonder if there is a way around this or at least make it a little less footgunny?

Reply:
Yeah it does seem weird that the required acl permissions could be updated separately, but we shouldn't need to modify the event at all outside of this endpoint. I don't think having to worry about calling SetACLRules will be an issue either - we only ever have to call that here in Subscribe on the server.